src/openrct2-ui/windows/Water.cpp

```diff
@@ -37,182 +37,148 @@ static rct_widget window_water_widgets[] = {
+    {
+        switch (widgetIndex)
+        {
+            case WIDX_DECREMENT:
+                // Decrement land tool size
+                gLandToolSize = std::max(MINIMUM_TOOL_SIZE, gLandToolSize - 1);
+
+                // Invalidate the window
+                Invalidate();
+                break;
+            case WIDX_INCREMENT:
+                // Increment land tool size
+                gLandToolSize = std::min(MAXIMUM_TOOL_SIZE, gLandToolSize + 1);
+
+                // Invalidate the window
+                Invalidate();
+                break;
+        }
+    }
 
-/**
- *
- *  rct2: 0x006E6BCE
- */
-static void window_water_update(rct_window* w)
-{
-    // Close window if another tool is open
-    if (!water_tool_is_active())
-        window_close(w);
-}
+    void OnUpdate() override
+    {
+        // Close window if another tool is open
+        if (!water_tool_is_active())
+        {
+            Close();
+        }
+    }
 
-/**
- *
- *  rct2: 0x006E6AB8
- */
-static void window_water_invalidate(rct_window* w)
-{
-    // Set the preview image button to be pressed down
-    w->pressed_widgets |= (1 << WIDX_PREVIEW);
+    void OnTextInput(rct_widgetindex widgetIndex, std::string_view text) override
+    {
+        int32_t size;
+        char* end;
+
+        if (widgetIndex != WIDX_PREVIEW)
+        {
+            return;
+        }
+
+        size = strtol(std::string(text).c_str(), &end, 10);
+        if (*end == '\0')
+        {
+            size = std::max(MINIMUM_TOOL_SIZE, size);
+            size = std::min(MAXIMUM_TOOL_SIZE, size);
+            gLandToolSize = size;
+
+            Invalidate();
+        }
+    }
 
-    // Update the preview image
-    window_water_widgets[WIDX_PREVIEW].image = LandTool::SizeToSpriteIndex(gLandToolSize);
-}
+    void OnPrepareDraw() override
+    {
+        // Set the preview image button to be pressed down
+        SetWidgetPressed(WIDX_PREVIEW, true);
 
-/**
- *
- *  rct2: 0x006E6ACF
- */
-static void window_water_paint(rct_window* w, rct_drawpixelinfo* dpi)
-{
-    auto screenCoords = ScreenCoordsXY{ w->windowPos.x + window_water_widgets[WIDX_PREVIEW].midX(),
-                                        w->windowPos.y + window_water_widgets[WIDX_PREVIEW].midY() };
+        // Update the preview image
+        widgets[WIDX_PREVIEW].image = LandTool::SizeToSpriteIndex(gLandToolSize);
+    }
 
-    WindowDrawWidgets(w, dpi);
-    // Draw number for tool sizes bigger than 7
-    if (gLandToolSize > MAX_TOOL_SIZE_WITH_SPRITE)
+    void OnDraw(rct_drawpixelinfo& dpi) override
     {
-        gfx_draw_string_centred(
-            dpi, STR_LAND_TOOL_SIZE_VALUE, screenCoords - ScreenCoordsXY{ 0, 2 }, COLOUR_BLACK, &gLandToolSize);
+        auto screenCoords = ScreenCoordsXY{ windowPos.x + window_water_widgets[WIDX_PREVIEW].midX(),
+                                            windowPos.y + window_water_widgets[WIDX_PREVIEW].midY() };
+
+        DrawWidgets(dpi);
```

Comment:
I believe this should be `DrawWidgets(&dpi);` with the change from *dpi to &dpi

Reply:
I just noticed this was changed from `WindowDrawWidgets()`. My comment is invalid.